src/mare/server.cr

```diff
@@ -85,6 +103,106 @@ class Mare::Server
+    # environment, the host's source path may not match ours, and we
+    # can apply the needed transformation as specified in this ENV var.
+    ENV["STD_DIRECTORY_MAPPING"]?.try do |mapping|
+      mapping.split(":").each_slice 2 do |pair|
+        next unless pair.size == 2
+        host_path = pair[0]
+        dest_path = pair[1]
+
+        if filename.starts_with?(host_path)
+          filename = 
+            if filename.includes?("prelude")
+              tmp_fname = filename.sub(host_path, Compiler.prelude_library_path)
+              tmp_fname.sub("prelude/prelude", "prelude")
+            else
+              filename.sub(host_path, Compiler::STANDARD_LIBRARY_DIRNAME)
+            end
+        end
+      end
+    end
+    ENV["SOURCE_DIRECTORY_MAPPING"]?.try do |mapping|
+      mapping.split(":").each_slice 2 do |pair|
+        next unless pair.size == 2
+        host_path = pair[0]
+        dest_path = pair[1]
+
+        if filename.starts_with?(host_path)
+          filename = filename.sub(host_path, dest_path)
+        end
+      end
+    end
+
+    dirname = File.dirname(filename)
+    sources = Compiler.get_library_sources(dirname)
+
+    source = sources.find { |s| s.path == filename }.not_nil!
+    source_pos = Source::Pos.point(source, pos.line.to_i32, pos.character.to_i32)
+
+    if @ctx.nil?
+      @ctx = Compiler.compile(sources, :serve_lsp)
+    end
+    ctx = @ctx.not_nil!
+
+    begin
+      definition_pos = ctx.serve_definition[source_pos]
+    rescue
+    end
+
+    if definition_pos.is_a? Mare::Source::Pos
+      user_filepath = definition_pos.source.path 
+
+      ENV["STD_DIRECTORY_MAPPING"]?.try do |mapping|
+        mapping.split(":").each_slice 2 do |pair|
+          next unless pair.size == 2
+          host_path = pair[0]
+          dest_path = pair[1]
+
+          if user_filepath.starts_with?(Compiler.prelude_library_path)
+            user_filepath = user_filepath.sub(Compiler.prelude_library_path, File.join(host_path, "prelude"))
+          end
+
+          if user_filepath.starts_with?(Compiler::STANDARD_LIBRARY_DIRNAME)
+            user_filepath = user_filepath.sub(Compiler::STANDARD_LIBRARY_DIRNAME, host_path)
+          end
+        end
+      end
+      ENV["SOURCE_DIRECTORY_MAPPING"]?.try do |mapping|
+        mapping.split(":").each_slice 2 do |pair|
+          next unless pair.size == 2
+          host_path = pair[0]
+          dest_path = pair[1]
+
+          if user_filepath.starts_with?(dest_path)
+            user_filepath = user_filepath.sub(dest_path, host_path)
+          end
+        end
+      end
+
+      @wire.respond msg do |msg|
+        msg.result = LSP::Data::Location.new(URI.new(path: user_filepath), LSP::Data::Range.new(LSP::Data::Position.new(definition_pos.row.to_i64, definition_pos.col.to_i64), LSP::Data::Position.new(definition_pos.row.to_i64, definition_pos.col.to_i64 + (definition_pos.finish - definition_pos.start))))
```

Comment:
Small formatting suggestion to break this large expression up over multiple lines:

```suggestion
        msg.result = LSP::Data::Location.new(
          URI.new(path: user_filepath),
          LSP::Data::Range.new(
            LSP::Data::Position.new(
              definition_pos.row.to_i64,
              definition_pos.col.to_i64,
            ),
            LSP::Data::Position.new(
              definition_pos.row.to_i64,
              definition_pos.col.to_i64 + (definition_pos.finish - definition_pos.start),
            ),
          ),
        )
```